crates/wasmtime/src/module/serialization.rs

```diff
@@ -276,45 +285,95 @@ impl<'a> SerializedModule<'a> {
         self.check_tunables(compiler)?;
         self.check_features(compiler)?;
 
-        let types = self
-            .tables
-            .into_iter()
-            .map(|t| Arc::new(t.unwrap_owned()))
-            .collect::<Vec<_>>();
-        let module = mk(engine, &types, self.data)?;
+        let types = Arc::new(self.types.unwrap_owned());
+        let mut modules = CompiledModule::from_artifacts_list(
+            self.artifacts
+                .into_iter()
+                .map(|i| i.unwrap_owned())
+                .collect(),
+            engine.compiler().isa(),
+            &*engine.config().profiler,
+        )?;
 
         // Validate the module can be used with the current allocator
-        engine.allocator().validate(module.inner.module.module())?;
+        engine
+            .allocator()
+            .validate(modules.last().unwrap().module())?;
+
+        let (signatures, trampolines) = engine.register_module_signatures(
```

Comment:
As module construction gets a little more complicated, could this be outlined in a new wasmtime-only constructor on `Module`? I'm imagining that ideally this function and `Module::from_binary` would end up in the same place which si the only place that actually literally constructs a `Module`